python/dgl/nn/pytorch/explain/subgraphxexplainer.py

```diff
@@ -0,0 +1,540 @@
+    r"""Monte Carlo Tree Search Node.
+
+    Parameters
+    ----------
+    nodes: list
+        The node ids of the graph that are associated with this tree node.
+    pruning_action: str
+        A representation of the pruning action used to get to this node.
+    c_puct: flaot
+        The hyper-parameter to encourage exploration while searching.
+    num_visit: int
+        Number times this node has been visited.
+    total_reward: float
+        the total reward for all node visits.
+    immediate_reward: float
+        Immediate reward for selecting this node (property score).
+    """
+
+    def __init__(
+        self,
+        nodes,
+        pruning_action,
+        c_puct=10.0,
+        num_visit=0,
+        total_reward=0.0,
+        immediate_reward=0.0,
+    ):
+        self.nodes = nodes
+        self.a = pruning_action
+        self.c_puct = c_puct
+        self.num_visit = num_visit
+        self.total_reward = total_reward
+        self.immediate_reward = immediate_reward
+        self.children = []
+
+    def average_reward(self):
+        r"""Get the average reward for multiple visits.
+
+        Returns
+        -------
+        float
+            Return the averaged reward for multiple visits.
+        """
+        return self.total_reward / self.num_visit if self.num_visit > 0 else 0
+
+    def action_selection_criteria(self, total_visit_count):
+        r"""Get the action selection criteria of node.
+
+        Parameters
+        ----------
+        total_visit_count: float
+            The total visiting counts for all possible actions of node.
+
+        Returns
+        -------
+        float
+            Returns the action selection criteria of node.
+        """
+        return (
+            self.c_puct
+            * self.immediate_reward
+            * math.sqrt(total_visit_count)
+            / (1 + self.num_visit)
+        )
+
+
+class SubgraphXExplainer(nn.Module):
+    r"""SubgraphXExplainer model from `SubgraphXExplainer: On Explainability of
+    Graph Neural Networks via Subgraph Explorations <https://arxiv.org/pdf/2102.05152.pdf>`.
+
+    It identifies subgraphs from the original graph  that play a
+    critical role in GNN-based graph classification.
+
+    To generate an explanation, by efficiently exploring different subgraphs with
+    Monte Carlo tree search. They  use Shapley values as a measure of subgraph importance
+
+    Parameters
+    ----------
+    model: nn.Module
+        The GNN model to explain.
+    hyperparam: float
+        The hyperparameter that encourages exploration.
+    pruning_action: str
+        Pruning action either "high2low" or "low2high" (refer to paper).
+    num_child_expand: int
+        Max number of children a tree node is allowed to have/expand to.
+    local_radius: int
+        Number of local radius to calculate :obj:`l_shapley`.
+    sample_num: int
+        Sampling time of monte carlo sampling approximation for
+        :obj:`mc_shapley`.
+    """
```

Comment:
Missing leaf threshold node number

Reply:
Added, it was defined in `explain_graph`, but I refactored it.